builtin/credential/jwt/path_cel_login.go

```diff
@@ -0,0 +1,286 @@
+				Callback: b.pathCelLogin,
+			},
+			logical.ResolveRoleOperation: &framework.PathOperation{
+				Callback: b.pathResolveCelRole,
+			},
+		},
+
+		HelpSynopsis:    pathCelLoginHelpSyn,
+		HelpDescription: pathCelLoginHelpDesc,
+	}
+}
+
+func (b *jwtAuthBackend) pathResolveCelRole(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	txRollback, err := logical.StartTxStorage(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	defer txRollback()
+
+	config, err := b.config(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+	if config == nil {
+		return logical.ErrorResponse("could not load configuration"), nil
+	}
+	celRole, resp, err := b.getCelRoleFromLoginRequest(config, ctx, req, d)
+	if resp != nil || err != nil {
+		return resp, err
+	}
+
+	if err := logical.EndTxStorage(ctx, req); err != nil {
+		return nil, err
+	}
+
+	return logical.ResolveRoleResponse(celRole.Name)
+}
+
+func (b *jwtAuthBackend) getCelRoleFromLoginRequest(config *jwtConfig, ctx context.Context, req *logical.Request, d *framework.FieldData) (*celRoleEntry, *logical.Response, error) {
+	roleName := d.Get("role").(string)
+	if roleName == "" {
+		roleName = config.DefaultRole
+	}
+	if roleName == "" {
+		return nil, logical.ErrorResponse("missing role"), nil
+	}
+
+	role, err := b.getCelRole(ctx, req.Storage, roleName)
+	if err != nil {
+		return nil, nil, err
+	}
+	if role == nil {
+		return nil, logical.ErrorResponse("role %q could not be found", roleName), nil
+	}
+
+	return role, nil, nil
+}
+
+func (b *jwtAuthBackend) pathCelLogin(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	txRollback, err := logical.StartTxStorage(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	defer txRollback()
+
+	config, err := b.config(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+	if config == nil {
+		return logical.ErrorResponse("could not load configuration"), nil
+	}
+
+	celRoleEntry, resp, err := b.getCelRoleFromLoginRequest(config, ctx, req, d)
+	if resp != nil || err != nil {
+		return resp, err
+	}
+
+	token := d.Get("jwt").(string)
+	if len(token) == 0 {
+		return logical.ErrorResponse("missing token"), nil
+	}
+
+	// Get the JWT validator based on the configured auth type
+	validator, err := b.jwtValidator(config)
+	if err != nil {
+		return logical.ErrorResponse("error configuring token validator: %s", err.Error()), nil
+	}
+
+	// Validate JWT supported algorithms if they've been provided. Otherwise,
+	// ensure that the signing algorithm is a member of the supported set.
+	signingAlgorithms := toAlg(config.JWTSupportedAlgs)
+	if len(signingAlgorithms) == 0 {
+		signingAlgorithms = []jwt.Alg{
+			jwt.RS256, jwt.RS384, jwt.RS512, jwt.ES256, jwt.ES384,
+			jwt.ES512, jwt.PS256, jwt.PS384, jwt.PS512, jwt.EdDSA,
+		}
+	}
+
+	// Set expected claims values to assert on the JWT
+	expected := jwt.Expected{
```

Comment:
Maybe the comment about `celEnv` is by design; it seems from here like audiences, subjects, claims, &c would have to be fully vetted by Cel. Perhaps with something like:

```cel
'aud' in claims ? 
    (
        claims["aud"] != "" &&
        SetPolicies(["example"]) &&
        SetUserClaim("aud")
    ) : false
```

In particular, I don't see a way to modify the environment, so we can't remove `claim["aud"]` dynamically, unless we add a helper function for that (`DeleteTokenClaim(...)` perhaps).

I'm not sure, tbh, what the best approach would be. The verification isn't exactly modular and returning a map type seems to require we know keys in advance (and can't necessarily incrementally build the map...).

Reply:
(cc @fatima2003)

In PKI, I think because the role system was already very modular, we opted to expose each of the role's helper functions in the Cel environment, letting role-like validation occur from cel via chaining -- e.g., `ValidateNames(csr.dns) && ValidateIP(...)` &c -- some more refactoring would make that a touch more explicit.

But I'll have to think more what the best operator experience would be here. It'd be nice not to have to enforce the entire role-based system here... but also have the option to use pieces of it, if desired.